controller/dataplane/konnect_extension.go

```diff
@@ -0,0 +1,123 @@
+package dataplane
+
+import (
+	"context"
+	"strings"
+
+	"github.com/samber/lo"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kong/gateway-operator/api/v1alpha1"
+	"github.com/kong/gateway-operator/api/v1beta1"
+	dputils "github.com/kong/gateway-operator/internal/utils/dataplane"
+	"github.com/kong/gateway-operator/pkg/consts"
+	k8sutils "github.com/kong/gateway-operator/pkg/utils/kubernetes"
+	k8sresources "github.com/kong/gateway-operator/pkg/utils/kubernetes/resources"
+)
+
+// applyDataPlaneKonnectExtension gets the DataPlane as argument, and in case it references a KonnectExtension, it
+// fetch the referenced extension and applies the necessary changes to the DataPlane spec.
+func applyDataPlaneKonnectExtension(ctx context.Context, cl client.Client, dataplane *v1beta1.DataPlane) error {
+	if len(dataplane.Spec.Extensions) == 0 {
+		return nil
+	}
+
+	for _, extensionRef := range dataplane.Spec.Extensions {
+		if extensionRef.Group != v1alpha1.SchemeGroupVersion.Group || extensionRef.Kind != "DataPlaneKonnectExtension" {
+			continue
+		}
+		namespace := dataplane.Namespace
+		if extensionRef.Namespace != nil {
+			namespace = *extensionRef.Namespace
```

Comment:
What was the final conclusion about handling cross-namespace references here? Should we allow them without explicit grants?

Reply:
good point. I blocked cross-namespace refs in f2b31eccbb9e20dcfc54358d9d8f41a3d2ee3fc1